docs/design/restic.rst

```diff
@@ -0,0 +1,158 @@
+=======================
+Restic-based data mover
+=======================
+
+.. admonition:: Enhancement status
+
+   Status: Proposed
+
+This is a proposal to add `Restic <https://restic.readthedocs.io/en/stable/>`_
+as an additional data mover within Scribe. Restic is a data backup utility that
+copies the data to an object store (among other options).
+
+While the main purpose of Scribe is to perform asynchronous data replication,
+there are some use cases that are more "backup oriented" but that don't require
+a full backup application (such as Velero). For example, some users may deploy
+and version control their application via GitOps techniques. These users may be
+looking for a simple method that allows preserving (off-cluster) snapshots of
+their storage so that it can be restored if necessary.
+
+Considerations
+==============
+
+The ReplicationSource and ReplicationDestination CRs of Scribe would correspond
+to the ``backup`` and ``restore`` operations, respectively, of Restic.
+Furthermore, there are repository maintenance operations that need to be
+addressed. For example, Restic manages the retention of old backups (via its
+``forget`` operation) as well as freeing objects that are no longer used (via
+its ``prune`` operation).
+
+While both Restic and Rclone read/write to object storage, their strengths are
+significantly different. The Rclone data mover is primarily designed for
+managing 1-to-many replication relationships, using the object store as an
+intermediary. On each sync, Rclone updates the object bucket to be identical to
+the current version of the source volume, making no attempt to preserve previous
+images. This works well for replication scenarios, but it may not be desirable
+when protection from accidental data deletion is desired. On the other hand,
+Restic is well suited for maintaining a series of historical versions in an
+efficient manner, but it is not designed for syncing data. The restore operation
+makes no allowance for small delta transfers.
+
+CRD for Restic mover
+====================
+
+In the normal case, the expected usage would be to have a ReplicationSource that
+controls the periodic backups of the data. It would use the same "common volume
+options" that Rsync and Rclone use to create a point-in-time image prior to
+copying the data.
+
+Backup
+------
+
+Given that in the normal case, only the ReplicationSource would be used, the
+repository maintenance options should be set there.
+
+.. code:: yaml
+
+   ---
+   apiVersion: scribe/v1alpha1
+   kind: ReplicationSource
+   metadata:
+     name: source
+     namespace: myns
+   spec:
+     sourcePVC: pvcname
+     trigger:
+       schedule: "0 * * * *"  # hourly backups
+     restic:
+       ### Standard volume options
+       # ReplicationSourceVolumeOptions
+
+       ### Restic-specific options
+       pruneIntervalDays:  # How often to prune the repository (*int)
```

Comment:
should this be it's own resource independent of our actual replicationsource?

Reply:
I'm assuming a 1:1 correspondence between restic repo and volume. Under that model, I don't think we'd gain anything by making a new CR.

Reply:
Adding to this... whether it should be 1:1 or not... By not sharing:
- It removes some configuration that would be needed to determine "which volume" within the repo is being accessed
- It removes some security concerns (can't restore somebody else's data)
- Forget/prune only affects (and ties up) this one volume
- Loses potential data savings via cross vol dedup